### PR TITLE
SONARJAVA-1037 RSPEC-1858 "toString()" should never be called on a String object

### PR DESCRIPTION
**Pull request by merciesa**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  public void visitNode(Tree tree) {
+    MethodInvocationTree methodTree = (MethodInvocationTree) tree;
+    if (STRING_TO_STRING.matches(methodTree)) {
+      ExpressionTree expressionTree = extractBaseExpression(((MemberSelectExpressionTree) methodTree.methodSelect()).expression());
+      if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+        addIssue(expressionTree, String.format("\"%s\" is already a string, there's no need to call \"toString()\" on it.", ((IdentifierTree) expressionTree).identifierToken().text()));
+      } else if (expressionTree.is(Tree.Kind.STRING_LITERAL)) {
+        addIssue(expressionTree, "there's no need to call \"toString()\" on a string literal.");
+      } else if (expressionTree.is(Tree.Kind.METHOD_INVOCATION)) {
+        addIssue(expressionTree, String.format("\"%s\" returns a string, there's no need to call \"toString()\".", extractName(((MethodInvocationTree) expressionTree).methodSelect())));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java]: :new: MINORSplit this 185 characters long line (which is greater than 180 authorized).[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  key = "S1858",
+  name = "\"toString()\" should never be called on a String object",
+  tags = {"clumsy", "pitfall"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.UNDERSTANDABILITY)
+@SqaleConstantRemediation("5min")
+public class StringToStringCheck extends SubscriptionBaseVisitor {
+
+  private final static MethodInvocationMatcher STRING_TO_STRING = MethodInvocationMatcher.create()
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java]: :new: MINORReorder the modifiers to comply with the Java Language Specification.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AModifiersOrderCheck)

File: java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    return ImmutableList.of(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree methodTree = (MethodInvocationTree) tree;
+    if (STRING_TO_STRING.matches(methodTree)) {
+      ExpressionTree expressionTree = extractBaseExpression(((MemberSelectExpressionTree) methodTree.methodSelect()).expression());
+      if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+        addIssue(expressionTree, String.format("\"%s\" is already a string, there's no need to call \"toString()\" on it.", ((IdentifierTree) expressionTree).identifierToken().text()));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java]: :new: MINORSplit this 185 characters long line (which is greater than 180 authorized).[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      } else if (expressionTree.is(Tree.Kind.STRING_LITERAL)) {
+        addIssue(expressionTree, "there's no need to call \"toString()\" on a string literal.");
+      } else if (expressionTree.is(Tree.Kind.METHOD_INVOCATION)) {
+        addIssue(expressionTree, String.format("\"%s\" returns a string, there's no need to call \"toString()\".", extractName(((MethodInvocationTree) expressionTree).methodSelect())));
+      } else if (expressionTree.is(Tree.Kind.ARRAY_ACCESS_EXPRESSION)) {
+        addIssue(expressionTree, String.format("\"%s\" is an array of strings, there's no need to call \"toString()\".", extractName(((ArrayAccessExpressionTree) expressionTree).expression())));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java]: :new: MINORSplit this 194 characters long line (which is greater than 180 authorized).[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 new issues:
* 4 minor

Details:
* [Split this 185 characters long line (which is greater than 180 authorized).](https://github.com/SonarSource/sonar-java/blob/e33eb2ad838edc347e4b5e3da936c46e7edb39ea/java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java#L67)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)
* [Reorder the modifiers to comply with the Java Language Specification.](https://github.com/SonarSource/sonar-java/blob/e33eb2ad838edc347e4b5e3da936c46e7edb39ea/java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java#L52)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AModifiersOrderCheck)
* [Split this 194 characters long line (which is greater than 180 authorized).](https://github.com/SonarSource/sonar-java/blob/e33eb2ad838edc347e4b5e3da936c46e7edb39ea/java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java#L73)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)
* [Split this 185 characters long line (which is greater than 180 authorized).](https://github.com/SonarSource/sonar-java/blob/e33eb2ad838edc347e4b5e3da936c46e7edb39ea/java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java#L71)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

File: java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+@Rule(
+  key = "S1858",
+  name = "\"toString()\" should never be called on a String object",
+  tags = {"clumsy", "pitfall"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.UNDERSTANDABILITY)
+@SqaleConstantRemediation("5min")
+public class StringToStringCheck extends SubscriptionBaseVisitor {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java]: You should probably use `AbstractMethodDetection`

File: java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      } else if (expressionTree.is(Tree.Kind.ARRAY_ACCESS_EXPRESSION)) {
+        addIssue(expressionTree, String.format("\"%s\" is an array of strings, there's no need to call \"toString()\".",
+          extractName(((ArrayAccessExpressionTree) expressionTree).expression())));
+      }
+    }
+  }
+
+  private ExpressionTree extractBaseExpression(ExpressionTree tree) {
+    ExpressionTree expressionTree = tree;
+    while (true) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/StringToStringCheck.java]: not found of a while(true) but it is a matter of styling so ok.

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

**Comment by benzonico** [comment]
LGTM

**Pull request closed without merging**